### PR TITLE
Add trigger for http actions to receive the status code

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix? 

Add a trigger to act on the http response status code from a request.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuraiton files to keep working)

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1049
  
# Test Environment

- [ ] ESP32
- [x] ESP8266
- [ ] Windows
- [x] Mac OS
- [ ] Linux

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
binary_sensor:
  - platform: gpio
    id: btn
    pin:
      number: GPIO0
      inverted: true
    on_press:
      - http_request.get:
          url: http://10.1.1.44:8123
          on_response:
            then:
              - logger.log:
                  format: 'Response status: %d'
                  args:
                    - status_code
```

# Explain your changes

This allows automations to be written based on the response code which is currently only output to the logs.

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by glmnet** [comment]
Looks good to me, seems to be forward compatible with a future handling of a response payload

cc @mvturnho

**Comment by jesserockz** [comment]
> seems to be forward compatible with a future handling of a response payload

Thats what I was aiming for, but was just going for something small and quick for now =)

File: esphome/components/http_request/http_request.cpp [hunk 1]
```diff
@@ -24,7 +24,7 @@ void HttpRequestComponent::set_url(std::string url) {
   this->client_.setReuse(true);
 }
 
-void HttpRequestComponent::send() {
+void HttpRequestComponent::send(std::vector<HttpRequestResponseTrigger *> response_triggers) {
```
Comment (OttoWinter) [inline esphome/components/http_request/http_request.cpp:27]:
```suggestion
void HttpRequestComponent::send(const std::vector<HttpRequestResponseTrigger *> &response_triggers) {
```

With the current version, the vector would be copied on each function call. By declaring it a `const` reference, only the reference is passed.

**Review by OttoWinter — COMMENTED** [review]
LGTM, thanks :+1:

File: esphome/components/http_request/__init__.py [hunk 2]
```diff
@@ -189,4 +197,9 @@ def http_request_action_to_code(config, action_id, template_arg, args):
         )
         cg.add(var.add_header(key, template_))
 
+    for conf in config.get(CONF_ON_RESPONSE, []):
+        trigger = cg.new_Pvariable(conf[CONF_TRIGGER_ID])
+        cg.add(var.register_response_trigger(trigger))
+        yield automation.build_automation(trigger, [(int, "x")], conf)
```
Comment (OttoWinter) [inline esphome/components/http_request/__init__.py:203]:
```suggestion
        yield automation.build_automation(trigger, [(int, "status_code")], conf)
```

I assume the plan is to attach the response payload as another variable to this `on_response` automation, right? If so naming the `status_code` using a long name would be better.

**Pull request merged**